sdk/go/pulumi/errors.go

```diff
@@ -0,0 +1,40 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulumi
+
+import (
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type InputPropertyErrorDetails struct {
+	PropertyPath string
+	Reason       string
+}
+
+func NewInputPropertiesError(message string, errors ...InputPropertyErrorDetails) error {
+	// Convert the errors to a slice of proto messages.
+	errorDetails := pulumirpc.InputPropertiesError{}
+	for _, e := range errors {
+		errorDetails.Errors = append(errorDetails.Errors, &pulumirpc.InputPropertiesError_PropertyError{
+			PropertyPath: e.PropertyPath,
+			Reason:       e.Reason,
+		})
+	}
+
+	s, _ := status.Newf(codes.InvalidArgument, "%s", message).WithDetails(&errorDetails)
```

Comment:
Related to the comment below, I feel like a fluent-style pattern has more appeal in Go due to the way errors work/the prevalence of `Errorf`-like helpers, etc. Maybe something like:

```go
type InputPropertiesError struct { ... }
type InputPropertyErrorDetails struct { ... }

func NewInputPropertyError(propertyPath string, reason string) *InputPropertiesError
func InputPropertyErrorf(propertyPath string, format string, args ...interface{}) *InputPropertiesError
func NewInputPropertiesError(message string, details ...InputPropertyErrorDetails) *InputPropertiesError
func InputPropertiesErrorf(format string, args ...interface{}) *InputPropertiesError

func (ipe *InputPropertiesError) Error() string
func (ipe *InputPropertiesError) Unwrap() string
func (ipe *InputPropertiesError) WithDetails(details ...InputPropertyErrorDetails) *InputPropertiesError
```

could work? But again happy to be shot down on this!

Reply:
This works for me!